lib/spack/spack/caches.py

```diff
@@ -59,24 +60,31 @@ def _fetch_cache():
 
 class MirrorCache(object):
     def __init__(self, root, skip_unstable_versions):
-        self.root = os.path.abspath(root)
+        self.root = spack.util.url.parse(root)
         self.skip_unstable_versions = skip_unstable_versions
 
     def store(self, fetcher, relative_dest):
         """Fetch and relocate the fetcher's target into our mirror cache."""
 
         # Note this will archive package sources even if they would not
         # normally be cached (e.g. the current tip of an hg/git branch)
-        dst = os.path.join(self.root, relative_dest)
-        mkdirp(os.path.dirname(dst))
+        dst = spack.util.url.join(self.root, relative_dest)
+        local_dst = spack.util.url.local_file_path(dst)
+        if local_dst:
+            mkdirp(os.path.dirname(local_dst))
```

Comment:
I agree this seems like a debt, but I'm not sure what the best thing to do is.

The root of the difference is the fact that there are no "directories" in `s3://`; everything is stored in a flat hierarchy.  For things like index generation and listing, we do a prefix search.  This is just one example of how key prefixes in S3 are treated like pseudo-folders.

Ideally, all url schemes would get the same treatment, but there's no way to do this without some kind of compromise.  We can't have "folders" everywhere, since there is no such thing with S3, and emulating how S3 treats object keys in `file://` urls would entail potentially `mkdirp`-ing many levels of directories on each write, which raises more questions and usability concerns.  For example, if a user makes a typo in a `file://` url, would it better for Spack to throw an error or blindly create potentially many levels of unintended directories in a user's filesystem?

As another example of how this subtle difference matters, consider two paths: `a/b/c.txt` and `a/b`.  If these were paths from a `file://` url, they couldn't coexist, because `a/b` cannot be both a directory and a file.  For `s3://` urls, this is perfectly fine.